tests/test_scatterwidgets.py

```diff
@@ -1,58 +1,404 @@
+    plot_widget.plot()
+
+    plot_widget.update_hover_highlight(x_data["vec"][0], y_data["vec"][0])
+    assert plot_widget.hovered_point.data[0][0] == x_data["vec"][0]
+    assert plot_widget.hovered_point.data[0][1] == y_data["vec"][0]
+
+    plot_widget.update_hover_highlight(-1, -1)
+    assert plot_widget.data_point_label.text() == "Value: N/A"
+    assert plot_widget.hovered_point.data.size == 0
+
+
+def test_clear_hover_highlight(plot_widget, prepare_disc_test_data):
+    """Test clearing of hover highlight."""
+    x_data, y_data, color_data, x_label, y_label, color_label = prepare_disc_test_data
+    plot_widget._onClick(x_data, y_data, color_data, x_label, y_label, color_label)
+    plot_widget.plot()
+
+    plot_widget.update_hover_highlight(x_data["vec"][0], y_data["vec"][0])
+    assert plot_widget.hovered_point.data[0][0] == x_data["vec"][0]
+    assert plot_widget.hovered_point.data[0][1] == y_data["vec"][0]
+
+    plot_widget.clear_hover_highlight()
+    assert plot_widget.hovered_point.data.size == 0
+
+
+def test_toggle_drawing_mode(plot_widget):
+    """Test toggling of drawing mode."""
+    plot_widget.toggle_drawing_mode()
+    assert plot_widget.drawing is True
+    assert plot_widget.rectangle is False
+
+    plot_widget.toggle_drawing_mode()
+    assert plot_widget.drawing is False
+
+
+def test_toggle_rectangle_mode(plot_widget):
+    """Test toggling of rectangle mode."""
+    plot_widget.toggle_rectangle_mode()
+    assert plot_widget.rectangle is True
+    assert plot_widget.drawing is False
+
+    plot_widget.toggle_rectangle_mode()
+    assert plot_widget.rectangle is False
+
+
+def plot_coords_to_viewport(plot_widget, x, y):
+    plot_pos = pg.Point(x, y)
+    view_pos = plot_widget.scatter_plot.vb.mapViewToScene(plot_pos)
+    return plot_widget.scatter_plot.scene().views()[0].mapFromScene(view_pos)
+
+
+def send_mouse_event(event_type, widget, pos, button=QtCore.Qt.LeftButton):
+    """Send a mouse event to a widget."""
+    # necessary because of https://bugreports.qt.io/browse/QTBUG-5232
+    event = QtGui.QMouseEvent(event_type, pos, button, button, QtCore.Qt.NoModifier)
+    QtWidgets.QApplication.sendEvent(widget, event)
+
+
+def test_roi_from_mouse_events(qtbot, plot_widget):
+    """Test mouse press and release events for drawing ROIs."""
+    plot_widget.toggle_drawing_mode()
+    assert plot_widget.drawing is True
+
+    plot_widget.scatter_plot.setRange(xRange=[0, 100], yRange=[0, 100])
+
+    # point coordinates in plot coordinates
+    point1 = [10, 10]
+    point2 = [70, 70]
+    point3 = [10, 80]
+
+    # Convert plot coordinates to viewport coordinates
+    viewport_pos1 = plot_coords_to_viewport(plot_widget, *point1)
+    viewport_pos2 = plot_coords_to_viewport(plot_widget, *point2)
+    viewport_pos3 = plot_coords_to_viewport(plot_widget, *point3)
+
+    view_widget = plot_widget.scatter_plot.scene().views()[0].viewport()
```

Comment:
what does `[0]` refers to? Could this cause problems?

Reply:
The [0] in plot_widget.scatter_plot.scene().views()[0].viewport() refers to the first QGraphicsView in the list of views associated with the QGraphicsScene. As in case of a single graph there is only one view associated with the scene, this will not cause problems.